u-config: wrap with pkg-config-wrapper and update to 0.33.3

marcin-serwin opened the pull request:
Wrap u-config with pkg-config-wrapper to set the PKG_CONFIG_PATH correctly in dev shells utilising it.

Don't use `Makefile` for building to avoid `pkg-config` dependency.



## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

marcin-serwin committed:
u-config: wrap with pkg-config-wrapper

This sets the PKG_CONFIG_PATH correctly in dev shells utilising it.

marcin-serwin committed:
u-config: rename to u-config-unwrapped

Sigmanificient commented:
CI seems to fail with that new ` u-config = callPackage ../build-support/pkg-config-wrapper` declaration. Not sure what should be done

Sigmanificient commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 357648`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>u-config</li>
    <li>u-config-unwrapped</li>
  </ul>
</details>

marcin-serwin committed:
u-config: manually invoke cc instead of using Makefile

The Makefile in the repo is not intended for packagers and recommends
manual compilation. This also lets us drop the `pkg-config` dependency.

r-ryantm committed:
u-config-unwrapped: 0.33.1 -> 0.33.3

marcin-serwin committed:
u-config-unwrapped: install man page

marcin-serwin committed:
u-config-unwrapped: add checks

marcin-serwin committed:
u-config-unwrapped: add update script

Sigmanificient reviewed (approved):
You seem more competent than me to maintain this package tbh :sweat_smile:

Sigmanificient commented on pkgs/by-name/u-/u-config-unwrapped/package.nix:
```diff
@@ -0,0 +1,62 @@
+
+  dontConfigure = true;
+
+  buildPhase = ''
+    runHook preBuild
+
+    foundMakefile=1 # for checkPhase
+    $CC -Os -o ${finalAttrs.meta.mainProgram} generic_main.c
+
+    runHook postBuild
+  '';
```
Is there a reason to not use the build Makefile?

Sigmanificient replied:
Maybe a comment about this could be nice

Sigmanificient commented on pkgs/by-name/u-/u-config-unwrapped/package.nix:
```diff
@@ -0,0 +1,62 @@
+    runHook preBuild
+
+    foundMakefile=1 # for checkPhase
+    $CC -Os -o ${finalAttrs.meta.mainProgram} generic_main.c
```
Not sure what the best to do tbh :sweat_smile:

marcin-serwin committed:
u-config-unwrapped: add marcin-serwin as a maintainer

N-R-K commented on pkgs/by-name/u-/u-config-unwrapped/package.nix:
```diff
@@ -0,0 +1,80 @@
+    # files are installed, hence the default Makefile does not contain target for building
+    # and installation, only for development. Instead of using it, we build the u-config via
+    # manual CC invocation. Since on NixOS there is no global pc-path we set it to /var/empty
+    # and rely on pkg-config-wrapper to provide the correct search path at runtime.
```
Just a heads up: `pc_path`, `pc_system_includedirs` & `pc_system_libdirs` virtual variables are expected to output the ["compiled in" paths and not change due to runtime configuration](https://github.com/skeeto/u-config/commit/ccd27ff688208c19bb2fbb8869876a295350c8af). Some build systems or whatnot may try to query it.

But I'm not a nix user though, so I cannot comment on whether this is fine or not. If it's fine to output `/var/empty[/...]` for these virtual variables then you can ignore this.

N-R-K replied:
> I was successful but only after applying [skeeto/u-config@62c88c0](https://github.com/skeeto/u-config/commit/62c88c0803f05ba33deb6151a20ca5d6b61598b6). I don't know enough about bootstrapping to understand this but similar patch is currently applied to `pkg-config` in nixpkgs and it [was causing some problems](https://github.com/NixOS/nixpkgs/issues/334195#issuecomment-2428781605). I see that it's reverted in `u-config` now.

The [gentoo ebuild for glib](https://github.com/gentoo/gentoo/blob/824f8f537280e5419d64edeeb1f1112ff7e3c4a7/dev-libs/glib/glib-2.84.0.ebuild#L281-L288) does some shenanigans with `PKG_CONFIG_*` vars to make it work without patching pkgconf/pkg-config. Maybe you could employ similar workaround in nixos's glib build too?

Sigmanificient commented:
I invite you to discuss over #396636
